File: Calculator_modularize.java
import java.util.*;

public class Calculator_modularize {

	public static void main(String[] arga) {
		Scanner sc = new Scanner(System.in);
		String line = sc.next();
		ArrayList<Token> tokens = tokenize(line);
		double answer = evaluate(tokens);
		System.out.println(line + "=" + answer);
	}

	static ArrayList<Token> tokenize(String line) {
		ArrayList<Token> tokens = new ArrayList<>();
		int lineIndex = 0;
		while (lineIndex < line.length()) {
			char c = line.charAt(lineIndex);
			if (isDigit(c))  lineIndex = readNumber(line, lineIndex, tokens);
			else if (c == '+') lineIndex = readPlus(line, lineIndex, tokens);
			else if (c == '-') lineIndex = readMinus(line, lineIndex, tokens);
			else if (c == '*') lineIndex = readMult(line, lineIndex, tokens);
			else if (c == '/') lineIndex = readDiv(line, lineIndex, tokens);
			else if (c == '(') lineIndex = readOpenParen(line, lineIndex, tokens);
			else if (c == ')') lineIndex = readCloseParen(line, lineIndex, tokens);
			else {
				System.out.println("Invalid character found:" + c);
				System.exit(1);
			}
		}
		return tokens;
	}

	static double evaluate(ArrayList<Token> tokens) {
		return evaluateParenExpr(tokens);
	}

	static double evaluateNoParenExpr(ArrayList<Token> tokensHasNoParens) {
		ArrayList<Token> tokensHasNo_MULT_DIV = mul_div(tokensHasNoParens);
		double answer = add_sub(tokensHasNo_MULT_DIV);
		Token token = new Token("NUMBER", answer);
		tokensHasNoParens.add(token);
		return answer;
	}

	static double evaluateParenExpr(ArrayList<Token> tokens) {
		int indexOfToken = 0;
		ArrayList<Token> tokensHasNoParens = new ArrayList<Token>();
		ArrayList<Token> tokensInsideParens = new ArrayList<>();
		while (indexOfToken < tokens.size()) {
			String type = tokens.get(indexOfToken).type;
			if (type.equals("OPEN_PAREN")) {
				int indexOfTokenInsideParens = indexOfToken+1;
				int parenCount = 1;
				while (indexOfToken < tokens.size()) {
					indexOfToken++;
					type = tokens.get(indexOfToken).type;
					if (type.equals("OPEN_PAREN")) parenCount++;
					if (type.equals("CLOSE_PAREN")) parenCount--;
					if (parenCount == 0) break;
					tokensInsideParens.add(tokens.get(indexOfToken));
				}
				double resultInside = evaluateParenExpr(tokensInsideParens);
				Token token = new Token("NUMBER", resultInside);
				tokensHasNoParens.add(token);
			} else {
				tokensHasNoParens.add(tokens.get(indexOfToken));
			}
			indexOfToken++;
		}
		return evaluateNoParenExpr(tokensHasNoParens);
	}

	static ArrayList<Token> mul_div(ArrayList<Token> tokens) {
		ArrayList<Token> tokensHasNo_MULT_DIV = new ArrayList<>(tokens);
		int index = 0;
		int removeIndex = 0;
		double result = 0;
		while (index < tokens.size()) {
			String type = tokens.get(index).type;
			if (type.equals("MULT") || type.equals("DIV")) {
				double x = tokens.get(index-1).value;
				double y = tokens.get(index+1).value;
				if (type.equals("MULT")) result = x * y;
				else if (type.equals("DIV")) result = x / y;
				Token token = new Token("NUMBER", result);
				removeIndex--;
				tokensHasNo_MULT_DIV.remove(removeIndex);
				tokensHasNo_MULT_DIV.remove(removeIndex);
				tokensHasNo_MULT_DIV.remove(removeIndex);
				tokensHasNo_MULT_DIV.add(removeIndex,token);
				tokens.remove(index+1);
				tokens.add(index+1,token);
				removeIndex++;
				index += 2;
			} else {
				index++;
				removeIndex++;
			}
		}
		return tokensHasNo_MULT_DIV;
	}

	static double add_sub(ArrayList<Token> tokens) {
		int index = 1;
		double answer = 0.0;
		if (tokens.get(0).type.equals("NUMBER")) {
			Token dummyPlus = new Token("PLUS", -1);
			tokens.add(0, dummyPlus);
		}
		while (index < tokens.size()) {
			if (tokens.get(index).type.equals("NUMBER")) {
				double value = tokens.get(index).value;
				String type = tokens.get(index-1).type;
				if (type.equals("PLUS")) answer += value;
				else if (type.equals("MINUS")) answer -= value;
				index++;
			}
			index++;
		}
		return answer;
	}
	static int readNumber(String line, int lineIndex, ArrayList<Token> tokens) {
		double num = 0.0;
		char c = line.charAt(lineIndex);
		while (lineIndex < line.length() && isDigit(c)) {
			num = num * 10 + changeToDouble(c);
			changeToDouble(c);
			lineIndex++;
			if (lineIndex >= line.length()) break;
 			c = line.charAt(lineIndex);
		}
		if (c == '.') {
			lineIndex++;
			double keta = 0;
			c = line.charAt(lineIndex);
			while (lineIndex < line.length() && isDigit(c)) {
				num = num*10 + changeToDouble(c);
				changeToDouble(c);
				keta++;
				lineIndex++;
				c = line.charAt(lineIndex);
			}
			num /= Math.pow(10, keta);
		}
		Token token = new Token("NUMBER", num);
		tokens.add(token);
		return lineIndex;
	}

	static int readPlus(String line, int lineIndex, ArrayList<Token> tokens) {
		Token token = new Token("PLUS", -1);
		tokens.add(token);
		lineIndex++;
		return lineIndex;
	}

	static int readMinus(String line, int lineIndex, ArrayList<Token> tokens) {
		Token token = new Token("MINUS", -1);
		tokens.add(token);
		lineIndex++;
		return lineIndex;
	}

	static int readMult(String line, int lineIndex, ArrayList<Token> tokens) {
		Token token = new Token("MULT", -1);
		tokens.add(token);
		lineIndex++;
		return lineIndex;
	}

	static int readDiv(String line, int lineIndex, ArrayList<Token> tokens) {
		Token token = new Token("DIV", -1);
		tokens.add(token);
		lineIndex++;
		return lineIndex;
	}

	static int readOpenParen(String line, int lineIndex, ArrayList<Token> tokens) {
		Token token = new Token("OPEN_PAREN", -1);
		tokens.add(token);
		lineIndex++;
		return lineIndex;
	}

	static int readCloseParen(String line, int lineIndex, ArrayList<Token> tokens) {
		Token token = new Token("CLOSE_PAREN", -1);
		tokens.add(token);
		lineIndex++;
		return lineIndex;
	}
	static boolean isDigit(char c) {
		return c >= '0' && c <= '9';
	}

	static double changeToDouble(char c) {
		return c - '0';
	}

	static class Token {
		String type;
		double value;
		public Token(String type, double value) {
			this.type = type;
			this.value = value;
		}
	}	
}

File: calculator_modularize.py
def readNumber(line, index):
  number = 0
  while index < len(line) and line[index].isdigit():
    number = number * 10 + int(line[index])
    index += 1
  if index < len(line) and line[index] == '.':
    index += 1
    keta = 0.1
    while index < len(line) and line[index].isdigit():
      number += int(line[index]) * keta
      keta /= 10
      index += 1
  token = {'type': 'NUMBER', 'number': number}
  return token, index


def readPlus(line, index):
  token = {'type': 'PLUS'}
  return token, index + 1

def readMinus(line, index):
  token = {'type': 'MINUS'}
  return token, index + 1

def readMult(line, index):
  token = {'type': 'MULT'}
  return token, index + 1

def readDiv(line, index):
  token = {'type': 'DIV'}
  return token, index + 1

def tokenize(line):
  tokens = []
  index = 0
  while index < len(line):
    if line[index].isdigit():
      (token, index) = readNumber(line, index)
    elif line[index] == '+':
      (token, index) = readPlus(line, index)
    elif line[index] == '-':
      (token, index) = readMinus(line, index)
    elif line[index] == '*':
      (token, index) = readMult(line, index)
    elif line[index] == '/':
      (token, index) = readDiv(line, index)
    else:
      print('Invalid character found: ' + line[index])
      exit(1)
    tokens.append(token)
  return tokens


def evaluate(tokens):
  answer = 0
  tokens.insert(0, {'type': 'PLUS'}) # Insert a dummy '+' token
  index = 1

  while index < len(tokens):
    temp_result = 0
    op = tokens[index-1]['type']
    #先に *, /　の処理
    if tokens[index]['type'] == 'NUMBER' and op != 'PLUS' and op != 'MINUS':
      y = tokens.pop(index)
      op = tokens.pop(index-1)
      x = tokens.pop(index-2)
      if x['type'] != 'NUMBER' or y['type'] != 'NUMBER':
        print('Wrong grammer') 
        exit(1)
      y = y['number']
      x = x['number']
      op = op['type']
      if op == 'MULT':
        temp_result = x * y
        token = {'type': 'NUMBER', 'number': temp_result}
        index = index-2
        tokens.insert(index, token)
      elif op == 'DIV':
        temp_result = x / y
        token = {'type': 'NUMBER', 'number': temp_result}
        index = index-2
        tokens.insert(index, token)
    index += 1
  index = 1
  while index < len(tokens):
    if tokens[index]['type'] == 'NUMBER':
      if tokens[index - 1]['type'] == 'PLUS':
        answer += tokens[index]['number']
      elif tokens[index - 1]['type'] == 'MINUS':
        answer -= tokens[index]['number']
      else:
        print('Invalid syntax2')
        exit(1)
      if index-2 >= 0 and tokens[index-2]['type'] != 'NUMBER':
        print("Wrong grammer2")
        exit(1)
    index += 1
  return answer


def test(line):
  tokens = tokenize(line)
  actualAnswer = evaluate(tokens)
  expectedAnswer = eval(line)
  if abs(actualAnswer - expectedAnswer) < 1e-8:
    print("PASS! (%s = %f)" % (line, expectedAnswer))
  else:
    print("FAIL! (%s should be %f but was %f)" % (line, expectedAnswer, actualAnswer))


# Add more tests to this function :)
def runTest():
  print("==== Test started! ====")
  #op → +, -だけ
  #整数
  #整数と小数 + 項が2つより多いとき
  #答えがマイナス
  test("1+2") 
  test("1.0-2.534+300")
  test("23-500+4.3-9.0") 

  #op　→ *
  #基本
  #整数と小数 + 項が2つより多いとき
  #答えマイナス
  test("2*3")
  test("2*2*2.3")
  test("2*3*1.2")

  #op　→ /
  #基本
  #整数と小数 + 項が2つより多いとき
  #答えマイナス
  #- * - →　+
  test("9/3")
  test("4.5/3/15/2.0")
  #test("4.5/3/15/-2.0")
  #test("-4.5/3/15/-2.0")

  #op → +, -, * ,/
  #基本
  #小数有り + マイナス
  #マイナス*マイナスを含む
  test("1+4*4+1/2")
  #test("3.4*1.2/2-43+2")
  #test("-4.3/-1.4+1-1*2.0")

  #エラー
  #test("1++2")
  test("1*2//2")
  print("==== Test finished! ====\n")

runTest()

while True:
  print('> ', end="")
  line = input()
  tokens = tokenize(line)
  answer = evaluate(tokens)
  print("answer = %f\n" % answer)